### PR TITLE
Feedback in navigation2 actions

### PR DESCRIPTION
**Pull request by SteveMacenski**
---

## Basic Info

| Info | Please fill out this column |
| ------ | ----------- |
| Ticket(s) this addresses   | #1339  |
| Primary OS tested on | Ubuntu |
| Robotic platform tested on | Gazebo |

---

## Description of contribution in a few bullet points

* Added feedback to long running actions.
* removed random crawl as it is still in nav2 experimental with the server

* NavigateToPose; pose, time spent on task, number of recoveries executed, distance remaining
* Follow path: Instantaneous speed, distance to goal in odom
* Spin and backup: distance so far gone
* Wait and compute path: none, as either has a while loop naturally conducive to feedback. 

## Description of documentation updates required from your changes

* **TODO** add to migration guide for Foxy

---

## Future work that may be required in bullet points

* **TODO** test each metric is properly computed and reset between runs.

This PR is also to gather reviews and feedback on other metrics you'd like to have included in the API.

### CODE REVIEW

File: nav2_controller/src/nav2_controller.cpp [hunk 1]
```diff
@@ -321,6 +322,13 @@ void ControllerServer::computeAndPublishVelocity()
     pose,
     nav_2d_utils::twist2Dto3D(twist));
 
+  std::shared_ptr<Action::Feedback> feedback = std::make_shared<Action::Feedback>();
+  feedback->speed = cmd_vel_2d.twist.linear.x;
```
Comment (cdelsey) [inline nav2_controller/src/nav2_controller.cpp]:
```suggestion
  feedback->speed = std::hypot(cmd_vel_2d.twist.linear.x, cmd_vel_2d.twist.linear.y);
```
To support holonomic robots.

File: nav2_controller/src/nav2_controller.cpp [hunk 2]
```diff
@@ -321,6 +322,13 @@ void ControllerServer::computeAndPublishVelocity()
...
+  feedback->distance_to_goal = sqrt(
```
Comment (cdelsey) [inline nav2_controller/src/nav2_controller.cpp]: You could also use std::hypot here. It's slower, but I don't think that matters in this case and results in less code.

File: nav2_msgs/action/FollowPath.action [hunk 3]
```diff
@@ -6,3 +6,5 @@ string controller_id
 std_msgs/Empty result
 ---
 #feedback
+float32 distance_to_goal
+float32 speed
```
Comment (cdelsey) [inline nav2_msgs/action/FollowPath.action]: I feel like the current distance from the path could be useful feedback, especially in implementations that use a more control theory oriented implementation like MPC.

File: nav2_behavior_tree/plugins/action/back_up_action.cpp [hunk 4]
```diff
@@ -51,6 +51,14 @@ class BackUpAction : public BtActionNode<nav2_msgs::action::BackUp>
     goal_.speed = speed;
   }
 
+  void on_tick() override
+  {
+    int recovery_count = 0;
```
Comment (cdelsey) [inline nav2_behavior_tree/plugins/action/back_up_action.cpp]: All this recovery count stuff would be good to push to a base class or helper function so it doesn't need to be repeated.

File: nav2_msgs/action/FollowPath.action (hunk 3 above)
Comment (SteveMacenski) [reply nav2_msgs/action/FollowPath.action]: Perhaps, but that is non-trivial to compute and I don't want to slow down the controller process too much. It'll involve at least a few vector multiplications and a cubic spline.

File: nav2_behavior_tree/plugins/action/back_up_action.cpp (hunk 4 above)
Comment (SteveMacenski) [reply nav2_behavior_tree/plugins/action/back_up_action.cpp]: Good point

**Pull request closed without merging**